duckdb-vortex/src/vortex_scan.cpp

```diff
@@ -349,63 +408,46 @@ void RegisterVortexScanFunction(DatabaseInstance &instance) {
 
 	vortex_scan.init_global = [](ClientContext &context,
 	                             TableFunctionInitInput &input) -> unique_ptr<GlobalTableFunctionState> {
-		auto &bind = input.bind_data->Cast<VortexBindData>();
-		auto state = make_uniq<VortexScanGlobalState>();
-
-		state->filter = input.filters;
-		state->projection_ids.reserve(input.projection_ids.size());
-		for (auto proj_id : input.projection_ids) {
-			state->projection_ids.push_back(input.column_ids[proj_id]);
-		}
-
-		state->column_ids = input.column_ids;
+		auto &bind = input.bind_data->CastNoConst<VortexBindData>();
+		auto global_state = make_uniq<VortexScanGlobalState>();
 
 		// TODO(joe): do this expansion gradually in the scan to avoid a slower start.
-		state->expanded_files = bind.file_list->GetAllFiles();
+		global_state->expanded_files = bind.file_list->GetAllFiles();
+		global_state->filter = input.filters;
+		global_state->column_ids = input.column_ids;
+
+		PopulateProjection(*global_state, bind.column_names, input);
 
 		// Most expressions are extracted from `PushdownComplexFilter`, the final filters come from `input.filters`.
-		vector<vortex::expr::Expr *> conjuncts;
-		std::copy(bind.conjuncts.begin(), bind.conjuncts.end(), std::back_inserter(conjuncts));
-		CreateFilterExpression(*bind.arena, bind.column_names, input.filters, input.column_ids, conjuncts);
-
-		auto column_names = std::vector<char const *>();
-		for (auto col_id : state->projection_ids) {
-			assert(col_id < bind.column_names.size());
-			column_names.push_back(bind.column_names[col_id].c_str());
+		CreateFilterExpression(*bind.arena, bind.column_names, input.filters, input.column_ids, bind.conjuncts);
+		if (auto exprs = flatten_exprs(*bind.arena, bind.conjuncts); exprs != nullptr) {
+			global_state->filter_str = exprs->SerializeAsString();
 		}
-		state->projected_column_names = column_names;
 
-		auto exprs = flatten_exprs(*bind.arena, bind.conjuncts);
-		if (exprs != nullptr) {
-			state->filter_str = exprs->SerializeAsString();
-		}
+		// Resizing the empty vector default constructs std::shared pointers at all indices with nullptr.
+		global_state->layout_readers.resize(global_state->expanded_files.size());
+		global_state->layout_mutexes = std::vector<std::mutex>(global_state->expanded_files.size());
 
-		// Can ignore mutex since no other threads are running now.
-		state->file_slots[0].array_stream = OpenArrayStream(bind, *state, bind.initial_file.get());
-		state->next_file = 1;
+		CreateScanPartitions(context, bind, *global_state);
```

Comment:
Not sure you can do this ahead of time, its a lot of work before any progress is made.

Reply:
This needs to be lazy in some way

Reply:
Hmm, maybe. I could spin off a worker thread here which pushes partitions into a lockfree queue.

Reply:
Preference though for putting this in a follow up PR.